To CPP or not to CPP...

On MacOS, defining the C preprocessor to be  `cpp` appears to be a sub-optimal choice. When invoked this way it runs with `-traditional-cpp`. One feature of that, apparently, is that the format of pre-processor directives is somewhat constrained. For example, no indentation. 

I see an unwelcome result of this when compiling libpng:

```
In file included from /usr/include/Availability.h:190:
/usr/include/AvailabilityInternal.h:22938:10: error: unterminated conditional directive
        #if __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_10_12
         ^
/usr/include/AvailabilityInternal.h:22867:10: error: unterminated conditional directive
        #if __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_10_11_4
         ^
... about 25 similar lines snipped ...

/usr/include/AvailabilityInternal.h:30:2: error: unterminated conditional directive
#ifndef __AVAILABILITY_INTERNAL__
 ^
In file included from ./png.h:397:
In file included from /usr/include/zlib.h:34:
In file included from /usr/include/zconf.h:439:
In file included from /usr/include/unistd.h:72:
In file included from /usr/include/sys/unistd.h:182:
/usr/include/Availability.h:239:2: error: #else without #if
#else
 ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```
Despite the errors, the compilation of libpng does complete.

I observe that the variables `$GCC`, `$GPP`, and `$GXX`, and their use in various `./configure` stanzas, were introduced into the `buildme.sh` script as a by-product of this commit: _Upgrade ICU to 58.2, necessary collation and build patches._  https://github.com/Logitech/slimserver-vendor/commit/cb7e0ad0e28d685c1a699f5aad9e92bfb1347807

I think they were introduced to support some differing FreeBSD build environments.

On MacOS, it seems that either `cc -E` or `gcc -E` are appropriate ways of invoking the preprocessor. I suspect that, left to their own devices,  the various configure scripts would work this out for themselves. And, possibly, the c-compiler and c++-compiler too ?

So, I think something should be changed, but I'm not clear what.  Could @fsbruva comment ?

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -99,8 +99,21 @@ if [ "$OS" != "Linux" -a "$OS" != "Darwin" -a "$OS" != "FreeBSD" -a "$OS" != "Su
 fi
 
 # Set default values prior to potential overwrite
-GCC=gcc
-GXX=g++
+# Check to see if CC and CXX are already defined
+if [[ ! -z "$CC" ]]; then
+   GCC="$CC"
+else
+   # Take a guess
+   GCC=gcc
+fi
+if [[ ! -z "$CXX" ]]; then
+   GXX="$CXX"
+else
+   GXX=g++
+fi
+
+# FreeBSD uses the following precedence: 1. Environment values for CC/CXX/CPP,
+# 2. Values defined in or 3. Stock build chain.
 if [ "$OS" = "FreeBSD" ]; then
     BSD_MAJOR_VER=`uname -r | sed 's/\..*//g'`
     BSD_MINOR_VER=`uname -r | sed 's/.*\.//g'`
@@ -109,7 +122,9 @@ if [ "$OS" = "FreeBSD" ]; then
         MAKE_CXX=`grep ^CXX= /etc/make.conf | grep -v CCACHE | grep -v \# | sed 's#CXX=##g'`
         MAKE_CPP=`grep ^CPP= /etc/make.conf | grep -v CCACHE | grep -v \# | sed 's#CPP=##g'`
     fi
-    if [[ ! -z "$MAKE_CC" ]]; then
+    if [[ ! -z "$CC" ]]; then
+        GCC="$CC"
+    elif [[ ! -z "$MAKE_CC" ]]; then
         GCC="$MAKE_CC"
     elif [ $BSD_MAJOR_VER -ge 10 ]; then
         # FreeBSD started using clang as the default compiler starting with 10.
@@ -117,7 +132,9 @@ if [ "$OS" = "FreeBSD" ]; then
     else
         GCC=gcc
     fi
-    if [[ ! -z "$MAKE_CXX" ]]; then
+    if [[ ! -z "$CXX" ]]; then
+        GXX="$CXX"
+    elif [[ ! -z "$MAKE_CXX" ]]; then
         GXX="$MAKE_CXX"
     elif [ $BSD_MAJOR_VER -ge 10 ]; then
         # FreeBSD started using clang++ as the default compiler starting with 10.
@@ -125,7 +142,9 @@ if [ "$OS" = "FreeBSD" ]; then
     else
         GXX=g++
     fi
-    if [[ ! -z "$MAKE_CPP" ]]; then
+    if [[ ! -z "$CPP" ]]; then
+        GPP="$CPP"
+    elif [[ ! -z "$MAKE_CPP" ]]; then
         GPP="$MAKE_CPP"
     else
         GPP=cpp

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -112,8 +112,10 @@ else
    GXX=g++
 fi
 
-# FreeBSD uses the following precedence: 1. Environment values for CC/CXX/CPP,
-# 2. Values defined in or 3. Stock build chain.
+# This script uses the following precedence for FreeBSD:
+# 1. Environment values for CC/CXX/CPP (checks if $CC is already defined)
+# 2. Values defined in /etc/make.conf, or
+# 3. Stock build chain
 if [ "$OS" = "FreeBSD" ]; then
     BSD_MAJOR_VER=`uname -r | sed 's/\..*//g'`
     BSD_MINOR_VER=`uname -r | sed 's/.*\.//g'`

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -267,7 +267,7 @@ if [ "$OS" = "FreeBSD" ]; then
 	    fi
 	done
 	for hdr in "zlib.h"; do
-	    hdr_found=$(find /usr/include -name "$hdr");
+	    hdr_found=$(find /usr/include/ -name "$hdr");
 	    if [ ! "$hdr_found" ]; then
 	        echo "$hdr not found - please install appropriate development package"
 	        exit 1

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -1395,12 +1395,16 @@ function build_ffmpeg {
 
     # ASM doesn't work right on x86_64
     # XXX test --arch options on Linux
-    if [ "$ARCH" = "x86_64-linux-thread-multi" -o "$ARCH" = "amd64-freebsd-thread-multi" -o "$ARCH" = "i86pc-solaris-thread-multi-64int" ]; then
+    if [[ "$ARCH" = "x86_64-linux-thread-multi" || "$ARCH" =~ "amd64-freebsd" || "$ARCH" = "i86pc-solaris-thread-multi-64int" ]]; then
         FFOPTS="$FFOPTS --disable-mmx"
     fi
     # FreeBSD amd64 needs arch option
-    if [ "$ARCH" = "amd64-freebsd" -o "$ARCH" = "amd64-freebsd-thread-multi" ]; then
+    if [[ "$ARCH" =~ "amd64-freebsd" ]]; then
         FFOPTS="$FFOPTS --arch=x86"
+        # FFMPEG has known issues with GCC 4.2. See: https://trac.ffmpeg.org/ticket/3970
+        if [[ "$CC_IS_GCC" == true && "$CC_VERSION" -ge 40200 && "$CC_VERSION" -lt 40300 ]]; then
+            FFOPTS="$FFOPTS --disable-asm"
+        fi
     fi
 
     if [ "$OS" = "Darwin" ]; then

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -795,7 +795,12 @@ function build {
             ;;
 
         Encode::Detect)
-            build_module Data-Dump-1.19
+            if [[ "$OS" == "FreeBSD" && `sysctl -n security.jail.jailed` == 1 && $PERL_MINOR_VER -le 10 ]]; then
+                # Tests fail in jails with old Perl
+                build_module Data-Dump-1.19 "" 0
+            else
+                build_module Data-Dump-1.19
+            fi
             build_module ExtUtils-CBuilder-0.260301
             build_module Module-Build-0.35 "" 0
             build_module Encode-Detect-1.00

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -664,6 +664,8 @@ function build {
         DBI)
             if [ $PERL_MINOR_VER -ge 18 ]; then
                 build_module DBI-1.628
+            elif [ $PERL_MINOR_VER -eq 8 ]; then
+                build_module DBI-1.616 "" 0
             else
                 build_module DBI-1.616
             fi

--- a/CPAN/buildme.sh
+++ b/CPAN/buildme.sh
@@ -98,10 +98,9 @@ if [ "$OS" != "Linux" -a "$OS" != "Darwin" -a "$OS" != "FreeBSD" -a "$OS" != "Su
     exit
 fi
 
-# Set default values prior to potential overwrite for FreeBSD platforms
+# Set default values prior to potential overwrite
 GCC=gcc
 GXX=g++
-GPP=cpp
 if [ "$OS" = "FreeBSD" ]; then
     BSD_MAJOR_VER=`uname -r | sed 's/\..*//g'`
     BSD_MINOR_VER=`uname -r | sed 's/.*\.//g'`
@@ -131,9 +130,14 @@ if [ "$OS" = "FreeBSD" ]; then
     else
         GPP=cpp
     fi
+    # Ensure the environment makes use of the desired/specified compilers and
+    # pre-processor
+    export CC=$GCC
+    export CXX=$GXX
+    export CPP=$GPP
 fi
 
-for i in $GCC $GPP rsync make ; do
+for i in $GCC $GXX rsync make ; do
     which $i > /dev/null
     if [ $? -ne 0 ] ; then
         echo "$i not found - please install it"
@@ -674,7 +678,6 @@ function build {
                     for i in ../../icu58_patches/freebsd/patch-*;
                         do patch -p0 < $i; done
                 fi
-                CC="$GCC" CXX="$GXX" CPP="$GPP" \
                 CFLAGS="$ICUFLAGS" CXXFLAGS="$ICUFLAGS" LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" \
                     ./runConfigureICU $ICUOS --prefix=$BUILD --enable-static --with-data-packaging=archive
                 $MAKE
@@ -939,7 +942,6 @@ function build {
             cd expat-2.0.1/conftools
             . ../../update-config.sh
             cd ..
-            CC="$GCC" CXX="$GXX" CPP="$GPP" \
             CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" \
             LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" \
                 ./configure --prefix=$BUILD \
@@ -986,7 +988,6 @@ function build {
             # libfreetype.a size (i386/x86_64 universal binary):
             #   1634288 (default)
             #    461984 (with custom ftoption.h/modules.cfg)
-            CC="$GCC" CXX="$GXX" CPP="$GPP" \
             CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" \
             LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" \
                 ./configure --prefix=$BUILD
@@ -1046,7 +1047,6 @@ function build {
             fi
             . ../update-config.sh
 
-            CC="$GCC" CXX="$GXX" CPP="$GPP" \
             CFLAGS="-I$BUILD/include $FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
             LDFLAGS="-L$BUILD/lib $FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
             OBJCFLAGS="-L$BUILD/lib $FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
@@ -1107,7 +1107,6 @@ function build_libexif {
     cd libexif-0.6.20
     . ../update-config.sh
 
-    CC="$GCC" CXX="$GXX" CPP="$GPP" \
     CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
     LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
         ./configure --prefix=$BUILD \
@@ -1233,7 +1232,6 @@ function build_libjpeg {
         # Disable features we don't need
         cp -fv ../libjpeg-turbo-jmorecfg.h jmorecfg.h
 
-        CC="$GCC" CXX="$GXX" CPP="$GPP" \
         CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" CXXFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS" \
             ./configure --prefix=$BUILD --disable-dependency-tracking
         $MAKE
@@ -1253,7 +1251,6 @@ function build_libjpeg {
         # Disable features we don't need
         cp -fv ../libjpeg-jmorecfg.h jmorecfg.h
 
-        CC="$GCC" CXX="$GXX" CPP="$GPP" \
         CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
         LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
             ./configure --prefix=$BUILD \
@@ -1285,7 +1282,6 @@ function build_libpng {
     cp -fv ../libpng-pngconf.h pngconf.h
     . ../update-config.sh
 
-    CC="$GCC" CXX="$GXX" CPP="$GPP" \
     CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
     LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
         ./configure --prefix=$BUILD \
@@ -1310,7 +1306,6 @@ function build_giflib {
     tar_wrapper zxf giflib-4.1.6.tar.gz
     cd giflib-4.1.6
     . ../update-config.sh
-    CC="$GCC" CXX="$GXX" CPP="$GPP" \
     CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
     LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
         ./configure --prefix=$BUILD \
@@ -1465,7 +1460,6 @@ function build_ffmpeg {
         FLAGS=$SAVED_FLAGS
         cd ..
     else
-        CC="$GCC" CXX="$GXX" CPP="$GPP" \
         CFLAGS="$FLAGS -O3" \
         LDFLAGS="$FLAGS -O3" \
             ./configure $FFOPTS
@@ -1504,7 +1498,6 @@ function build_bdb {
        patch -p0 < ../db51-src_dbinc_atomic.patch
        popd
     fi
-    CC="$GCC" CXX="$GXX" CPP="$GPP" \
     CFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
     LDFLAGS="$FLAGS $OSX_ARCH $OSX_FLAGS -O3" \
         ../dist/configure --prefix=$BUILD $MUTEX \

--- a/CPAN/icu58_patches/digitlst.cpp.patch
+++ b/CPAN/icu58_patches/digitlst.cpp.patch
@@ -39,3 +39,33 @@
  fi
  AC_SUBST(U_HAVE_STRTOD_L)
  
+--- configure.orig	2016-10-04 19:30:20.000000000 +0000
++++ configure	2018-06-24 01:04:40.476591000 +0000
+@@ -7053,11 +7053,23 @@
+ 
+ if test x$ac_cv_func_strtod_l = xyes
+ then
+-     CONFIG_CPPFLAGS="${CONFIG_CPPFLAGS} -DU_HAVE_STRTOD_L=1"
+-     U_HAVE_STRTOD_L=1
++    U_HAVE_STRTOD_L=1
++    ac_fn_c_check_header_mongrel "$LINENO" "xlocale.h" "ac_cv_header_xlocale_h" "$ac_includes_default"
++if test "x$ac_cv_header_xlocale_h" = xyes; then :
++
++fi
++
++
++    if test "$ac_cv_header_xlocale_h" = yes; then
++      U_HAVE_XLOCALE_H=1
++      CONFIG_CPPFLAGS="${CONFIG_CPPFLAGS} -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1"
++    else
++      U_HAVE_XLOCALE_H=0
++      CONFIG_CPPFLAGS="${CONFIG_CPPFLAGS} -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=0"
++    fi
+ else
+-     CONFIG_CPPFLAGS="${CONFIG_CPPFLAGS} -DU_HAVE_STRTOD_L=0"
+-     U_HAVE_STRTOD_L=0
++    CONFIG_CPPFLAGS="${CONFIG_CPPFLAGS} -DU_HAVE_STRTOD_L=0"
++    U_HAVE_STRTOD_L=0
+ fi
+ 
+ 

